llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp

```diff
@@ -0,0 +1,430 @@
+//===-- ESIMDVecArgPass.cpp - lower Close To Metal (CM) constructs --------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Change in function parameter type from simd* to native llvm vector type for
```

Comment:
is it legal for clang's CG to change type of parameter from simd* to say <i16x16>*?

Reply:
> simd* to say *?
simd* to say "<i16x32>*"?

Reply:
if you're referring to __attribute__((ext_vector_type(N)) then we do use this to declare member of class simd. the sole data member of class simd is of vector type with this attribute applied. simd class has several other methods implemented.

Reply:
> is it legal for clang's CG to change type of parameter from simd* to say *?

If it's legal for LLVM pass it should be legal for any other compiler component. According to my understanding changing type at CodeGen should much easier to maintain than in LLVM pass. CodeGen lowers AST types into LLVM types, so it "changes" type already. According to my understanding changing LLVM type by pass is much more difficult to do.

> if you're referring to **attribute**((ext_vector_type(N)) then we do use this to declare member of class simd. the sole data member of class simd is of vector type with this attribute applied. simd class has several other methods implemented.

Why do we expect `llvm.genx.vload.v16i32.p4v16i32` intrinsic to accept simd class as a parameter instead of vector type data member? Could you point to the SYCL headers source code where this intrinsic is used, please? I'd like to check if it's possible to handle the problem in headers instead of LLVM pass.

Reply:
well, by legal i mean simd class has several operations defined on it which arent defined on underlying vector type. LLVM pass that is uploaded here assumes that replacing parent class with its sole member is safe only for the special case where a class has a single member that is a native vector type. within the callee, vector* is first bitcast to simd* so all existing users of the older simd* argument continue to work as before. optimizations ensure redundant casts are done away with and we end up with vector* types all around.

the reason we do this transformation is because backend vector compiler expects vector* type in its input in order to eventually promote them to GRF in final code.

here is definition of __vector_type - sycl\include\CL\sycl\intel\esimd\esimd_types.hpp
here is definition of simd class - sycl\include\CL\sycl\intel\esimd\esimd.hpp
here is a header defining memory access library functions - sycl\include\CL\sycl\intel\esimd\esimd_memory.hpp
intrinsics are lowered based on table in this file - llvm\lib\SYCLLowerIR\LowerESIMD.cpp

In esimd_memory.hpp, simd class is used as a parameter. this is a header that user is expected to use. however, internally it invokes methods from internal headers in esimd\detail that consume __vector_type_t.

my understanding is that llvm.genx.vload.v16i32.p4v16i32 takes vector argument and not simd argument.

Reply:
> Why do we expect llvm.genx.vload.v16i32.p4v16i32 intrinsic to accept simd class as a parameter instead of vector type data member? Could you point to the SYCL headers source code where this intrinsic is used, please? I'd like to check if it's possible to handle the problem in headers instead of LLVM pass.

@bader , llvm.genx* intrinsics don't know anything about the simd class, their parameters are native LLVM vectors. See for example esimd_memory.hpp:
```cpp
/// flat-address atomic, one source operand, add/sub/min/max etc
template <EsimdAtomicOpType Op, typename T, int n,
          CacheHint L1H = CacheHint::Default,
          CacheHint L3H = CacheHint::Default>
ESIMD_NODEBUG ESIMD_INLINE
    typename std::enable_if<check_atomic<Op, T, n, 1>(), simd<T, n>>::type
    flat_atomic(T *p, simd<unsigned, n> offset, simd<T, n> src0,
                simd<ushort, n> pred) {
  simd<uintptr_t, n> vAddr(reinterpret_cast<uintptr_t>(p));
  simd<uintptr_t, n> offset_i1 = convert<uintptr_t>(offset);
  vAddr += offset_i1;
  return __esimd_flat_atomic1<Op, T, n, L1H, L3H>(vAddr.data(), src0.data(),
                                                  pred.data());
}
```
here all simd objects are unwrapped via `.data()` when invoking the source-level CPP intrinsic `__esimd_flat_atomic1`, which is then translated by ESIMD lower to  `@llvm.genx.svm.atomic.`

This can't be done similar way for user functions because they are user-defined.
Also, I think the reason why this type translation is done not in the FE but in an LLVM pass is that not all occurrences of the type is translated that way - only actual and formal parameters. It is thus too ESIMD-specific.

Reply:
I don't understand the problem. I think you should put clear description of the problem to the PR description.
Why do we need change user function parameters? C++ is strongly typed language, so instead of hacking function parameters in LLVM passes, user should defined define them.

What is the problem with using simd class instead of LLVM vector types in user function?

E.g. 
```
// Translate simd* parameter (#1) to vector <16 x 32>* type and insert bitcast.
// All users of old parameter will use result of the bitcast.
```
This looks like a work-around for some back-end issue.
I'm quite sure that if we use X86 back-end, it will "generate correct code" and it will use vector ISA.

Reply:
I think X86 will still generate correct code as binary rep of simd and vector are the same. The BE issue is that it does not know about simd and needs known types in parameters to generate good code. @pratikashar, @cmc-rep, please comment.

Reply:
> // Translate simd* parameter (#1) to vector <16 x 32>* type and insert bitcast.
> // All users of old parameter will use result of the bitcast.
@bader - Maybe there is a misunderstanding in above statement. The idea is *all* functions, including user/library functions in LLVM IR, will have their simd* argument converted to vector*. Let me share an example in case there is a misunderstanding:

void bar(simd* arg1, simd* arg2)
{
}

void foo(simd* arg1)
{
   auto arg2 = arg1;
   bar(arg1, arg2);
}

==> This will get translated roughly to:

void bar(vector* arg1, vector* arg2)
{
}

void foo(vector* arg1)
{
    ; LLVM IR in prolog:
    ; arg1_simd = bitcast vector* arg1 to simd*
    auto arg2 = arg1_simd; <-- 
    ; arg2_v = bitcast vector* arg2 to vector*
    bar(arg1, arg2_v);
}

This guarantees that all functions in input to backend have vector* as function parameter types. Within function foo and bar, any uses of arg1 such as copy constructors or other simd calls method calls on arg1 would continue to work. Except they would now use arg1_simd as the base object rather than arg1.

The backend compiler that consumes SPIRV generated by dpcpp-cm compiler assumes all parameters that are supposed to be vectors/matrices have been converted to LLVM's native type. It's job is to take dpcpp-cm output and generate code with as many variables promoted to GRFs as possible for performance. Not adhering to assumption of backend compiler means it wont know how to promote argument vectors to GRF and may end up with memory accesses.

Gang can correct me if this is incorrect. But Cm's clang also emits LLVM native vectors as function parameters, which allows cmc compiler to generate good code. So I think esimd compiler should follow suit.

Reply:
> Not adhering to assumption of backend compiler means it wont know how to promote argument vectors to GRF and may end up with memory accesses.

That's what I don't understand.
Why the back-end compiler is able to able to promote arg1 to GRF here:
```
void foo(vector* arg1)
{
; LLVM IR in prolog:
; arg1_simd = bitcast vector* arg1 to simd*
auto arg2 = arg1_simd; <--
...
```
but not here:
```
void foo(simd* arg1)
{
auto arg2 = arg1;
```
As Konst and I noted in the comments above X86 backend should be able to generate identical code for both cases as LLVM type doesn't matter, but what matter is the size. sizeof(simd) == sizeof(vector), so I would expect the same assembly code for both types. Why is GPU back-end different?

So I don't understand why "simd" code is "incorrect" and "vector" code is "correct". 

Another concern is that LLVM pass is quite difficult to maintain. Any small changes to SYCL header might break this transformation and code become "incorrect".

Reply:
This lowering is required by the backend.  More work done in  offline compilation means less time spent in online compilation

Reply:
I don't think it requires any additional work impacting "compilation time".

Reply:
If we do not do this here, we have to do the same transformation in the backend, that is extra online compilation time.

Reply:
How do you decide which transformation should be done offline or online?

I still believe that doing this transformation 'here' is bad design. 
Alternative option:
- user change the source code to use `vector` type instead of `simd` to get better performance. (I think comments are not quite accurate and the code with simd is correct, but back-end is not able to optimize it)

Reply:
Restricting user from using simd type in function declaration will make the ESIMD extension less usable. Hence not a good choice to us.  The alternative of this pass is to change all SIMD type to vector type during clang CodeGen (somewhere in this long review, I see you indicated that is doable and better). You are the clang expert